hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java

```diff
@@ -1537,6 +1655,277 @@ public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
 		}
 	}
 
+	protected class AllowancePrecompile extends ERCReadOnlyAbstractPrecompile {
+		private TokenAllowanceWrapper allowanceWrapper;
+
+		public AllowancePrecompile(final TokenID tokenId) {
+			super(tokenId);
+		}
+
+		@Override
+		public TransactionBody.Builder body(final Bytes input, final UnaryOperator<byte[]> aliasResolver) {
+			final var nestedInput = input.slice(24);
+			allowanceWrapper = decoder.decodeTokenAllowance(nestedInput, aliasResolver);
+
+			return super.body(input, aliasResolver);
+		}
+
+		@Override
+		public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
+			final TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accountsLedger = ledgers.accounts();
+			final var allowance = (TreeMap<FcTokenAllowanceId, Long>) accountsLedger.get(allowanceWrapper.owner(), FUNGIBLE_TOKEN_ALLOWANCES);
+			long value = 0;
+			for (Map.Entry<FcTokenAllowanceId, Long> e : allowance.entrySet()) {
+				if (allowanceWrapper.spender().getAccountNum() == e.getKey().getSpenderNum().longValue() &&
+						tokenId.getTokenNum() == e.getKey().getTokenNum().longValue()) {
+					value = e.getValue();
+				}
+			}
+			return encoder.encodeAllowance(value);
+		}
+	}
+
+	protected class ApprovePrecompile implements Precompile {
+		protected ApproveWrapper approveOp;
+		protected TokenID token;
+		private final boolean isFungible;
+
+		public ApprovePrecompile(final TokenID token, final boolean isFungible) {
+			this.token = token;
+			this.isFungible = isFungible;
+		}
+
+		@Override
+		public TransactionBody.Builder body(final Bytes input, final UnaryOperator<byte[]> aliasResolver) {
+			final var accountId = EntityIdUtils.accountIdFromEvmAddress(senderAddress);
+			final var fungibleTokenAllowances = (Map<FcTokenAllowanceId, Long>) ledgers.accounts().get(accountId, AccountProperty.FUNGIBLE_TOKEN_ALLOWANCES);
+
+			final var nestedInput = input.slice(24);
+			approveOp = decoder.decodeTokenApprove(nestedInput, token, isFungible, aliasResolver);
+
+			if (isFungible) {
+				long value = 0;
+				for (Map.Entry<FcTokenAllowanceId, Long> e : fungibleTokenAllowances.entrySet()) {
+					if (approveOp.spender().getAccountNum() == e.getKey().getSpenderNum().longValue()) {
```

Comment:
Can't we just directly look up the right entry by constructing an `FcTokenAllowanceId` here from `token` and `approveOp.spender()`?